### ISSUE TITLE
Reticulated content behaves differently and returns incorrect results for PyTorch model on Linux

### ISSUE THREAD

**Issue by koverholt** [rstudio/reticulate#407]
### Versions:

R version: 3.5.1
Reticulate version: 1.10
Python version: 3.6.7

### Description:

When I run a Python script (image classifier with pytorch) via a Reticulated Rmd document, I get different results on my laptop vs. a Linux machine. On Mac, the model predictions are correct and consistent between runs. On Linux, the model predictions are random and incorrect. I can reproduce this random behavior using the steps below.

I was initially having this problem via a Reticulated Shiny app, but I transferred the code to an Rmd document and was able to reproduce the inconsistent behavior.

### Steps to reproduce:

Note that I configure the path to Python using the `use_python()` function in the Rmd example below whether I am running the document on Mac or Linux. On both the Mac and Linux machines, I am using the same versions of R, Python, and the same versions of the R and Python packages.

1) Create a Python environment on Mac machine with the packages listed in the `requirements.txt` below.
2) Run the below Rmd document on Mac and observe the resulting predictions, which remain the same if you re-run the document multiple times (see Expected Output). I used the following command:
   ```    
   R -e 'Sys.setenv(RSTUDIO_PANDOC="/Applications/RStudio.app/Contents/MacOS/pandoc"); rmarkdown::render("image-classifier.Rmd")'
   ```
3) Create a Python environment on Linux machine with the packages listed in the `requirements.txt` below.
4) Run the same Rmd document on Linux. I used the following command:
   ```    
   /opt/R/3.5.1/bin/R -e 'Sys.setenv(RSTUDIO_PANDOC="/usr/lib/rstudio/bin/pandoc"); rmarkdown::render("image-classifier.Rmd")'
   ```
5) Observe very different resulting predictions each time you re-run the Rmd document on Linux (see Actual Output).

### Expected output:

Since the image URL is fixed, the classification model (`squeezenet1_1`) is fixed/pre-trained and downloaded to the`./model/` dir upon execution, and the model is initialized in evaluation/test mode using `sqnet.eval()`, thus the model should always return the same predictions.

With these fixed parameters, the output I always get on Mac is as follows (even when executing multiple times):

```
## Egyptian cat 30.884262
## tiger cat 30.400385
## tabby, tabby cat 29.59031
## lynx, catamount 29.143005
## hen 24.640438
```

### Actual output:

When I run the Rmd document on Linux, I get random results for the categories and scores. For example, when I run the Rmd document three times, I get the following random categories and scores:

```
## ladle 224027.9
## wooden spoon 173143.23
## racket, racquet 167481.12
## hammer 163295.52
## letter opener, paper knife, paperknife 159362.58
```

```
## paintbrush 114514.62
## oil filter 107819.82
## candle, taper, wax light 105742.734
## measuring cup 105111.29
## goblet 103484.85
```

```
## tick 6850.0415
## green lizard, Lacerta viridis 6554.929
## leafhopper 6088.1333
## American chameleon, anole, Anolis carolinensis 5832.1206
## weevil 5398.471
```

Also note that the scores in the `predictions` array are many orders of magnitude higher than when I run the document on my Mac (see Expected Output).

### Files necessary to reproduce:

Contents of `image-classifier.Rmd`:

````
---
title: "Image classifier"
output: html_document
---

```{r setup}
library(reticulate)
use_python("/anaconda3/envs/image/bin/python3", required=TRUE)  # Used on Mac
# use_python("/opt/Python/3.6.7/bin/python3", required=TRUE)  # Used on Linux
```

```{python}
import io
import os
from PIL import Image
import requests
from torch.autograd import Variable
import torchvision.models as models
import torchvision.transforms as transforms

os.environ["TORCH_MODEL_ZOO"] = "./model"

IMG_URL = 'https://upload.wikimedia.org/wikipedia/commons/thumb/3/3a/Cat03.jpg/1200px-Cat03.jpg'
LABELS_URL = 'https://s3.amazonaws.com/outcome-blog/imagenet/labels.json'

response = requests.get(LABELS_URL)
labels = {int(key): value for key, value in response.json().items()}

response = requests.get(IMG_URL)
img = Image.open(io.BytesIO(response.content))

transform_pipeline = transforms.Compose([transforms.Resize(224),
                                         transforms.ToTensor(),
                                         transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                                              std=[0.229, 0.224, 0.225])])

img = transform_pipeline(img)
img = img.unsqueeze(0)
img = Variable(img)
sqnet = models.squeezenet1_1(pretrained=True)
sqnet.eval()
result = sqnet(img)
predictions = result.data.numpy()[0]
top_k = predictions.argsort()[-5:][::-1]
for ind in top_k:
    print(labels[ind], predictions[ind])
```

````

Contents of `requirements.txt`:

```
Pillow==5.2.0
requests==2.19.1
torch=1.0.0
torchvision==0.2.1
```

**Comment by kevinushey** [comment]
I just tried running your script on my Linux VM and all the predictions came out as expected. Perhaps there's something gone wrong with your R and/or Python installations? (I used the system Python 3.6 installation on Ubuntu 18.04)

**Comment by koverholt** [comment]
Interesting, thank you for testing. I am using Ubuntu 14.04.

The odd thing is that the predictions are correct if I execute a Python script with the same content directly, but the results are not correct when I execute the same code in the Rmd document with reticulate.

**Comment by kevinushey** [comment]
How did you install Python 3.6 on your system? I wonder if there's some kind of unexpected crosstalk between that version of Python and e.g. either another version of Python on the system, or perhaps wrong system libraries are being loaded.

It could be useful to see the output of `py_config()` after loading Python.

**Comment by koverholt** [comment]
I tried again on CentOS 7 with R 3.5.1 and Python 3.6.7 (installed the same way as the Ubuntu 14.04 machine), and things are working as expected.

I am at my limits of digging into this, but it seems like the issue could be related to outdated system libraries on Ubuntu 14.04, since you verified it works on Ubuntu 18.04 and I verified on CentOS 7.

And separately, when I try testing this with `pytorch` installed from conda on the Ubuntu 14.04 system, it complains about `ImportError: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /home/kris/anaconda/lib/python3.6/site-packages/torch/lib/libshm.so)`, which is why I think this is related to system libraries, but would require a dist-upgrade from 14.04 to get that version of glibcxx.

I'm going to close this issue since it is not strictly an issue with reticulate, but rather a combination of Python/system libraries available on Ubuntu 14.04 that happens to manifest with reticulate. Thanks for your help troubleshooting!

**Comment by johanneswaage** [comment]
I'm seeing the same behaviour on RHEL 7. 
I've created a virtual environment using 
`renv::use_python(python = "/opt/python/3.7.6/bin/python3.7")`

This version of python has been built from source using:

`./configure --prefix=/opt/python/3.7.6 --with-ensurepip=install --enable-shared --enable-optimizations
make -j 32
sudo make install
`

This creates a virtual environment in the project folder. 
I activate this environment and use pip to install the necessary dependencies.

Here's a print of `reticulate::py_config()`
`python:         /data/analyse_og_indsigt/users/jwa/repos/screenduck/renv/python/virtualenvs/renv-python-3.7.6/bin/python
libpython:      /opt/python/3.7.6/lib/libpython3.7m.so
pythonhome:     /opt/python/3.7.6:/opt/python/3.7.6
version:        3.7.6 (default, Jan  6 2020, 11:37:34)  [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
numpy:          /data/analyse_og_indsigt/users/jwa/repos/screenduck/renv/python/virtualenvs/renv-python-3.7.6/lib/python3.7/site-packages/numpy
numpy_version:  1.18.1`

NOTE: Python version was forced by RETICULATE_PYTHON`

Everything seems to look fine.

If I run the python pytorch script from R using reticulate, or reticulate::repl, prediction results are random.

If I source the script from within the python virtual environment, results come out correct.

What could be the reason for this? Thanks a bunch.

`
─ Session info ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.5.1 (2018-07-02)
 os       Red Hat Enterprise Linux    
 system   x86_64, linux-gnu           
 ui       RStudio                     
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Europe/Copenhagen           
 date     2020-01-12                  

─ Packages ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 ! package     * version   date       lib source        
 P assertthat    0.2.1     2019-03-21 [?] RSPM (R 3.5.1)
 P cli           1.1.0     2019-03-19 [?] RSPM (R 3.5.1)
 P crayon        1.3.4     2017-09-16 [?] RSPM (R 3.5.1)
 P jsonlite      1.6       2018-12-07 [?] RSPM (R 3.5.1)
 P packrat       0.5.0     2018-11-14 [?] RSPM (R 3.5.1)
   png           0.1-7     2013-12-03 [1] RSPM (R 3.5.1)
 P rappdirs      0.3.1     2016-03-28 [?] RSPM (R 3.5.1)
 P Rcpp          1.0.2     2019-07-25 [?] RSPM (R 3.5.1)
   renv          0.7.0-125 2020-01-12 [1] local         
 P reticulate    1.14      2019-12-17 [?] RSPM (R 3.5.1)
 P rstudioapi    0.10      2019-03-19 [?] RSPM (R 3.5.1)
 P sessioninfo   1.1.1     2018-11-05 [?] RSPM (R 3.5.1)
 P withr         2.1.2     2018-03-15 [?] RSPM (R 3.5.1)

[1] /data/analyse_og_indsigt/users/jwa/repos/screenduck/renv/library/R-3.5/x86_64-pc-linux-gnu
[2] /tmp/Rtmp0FA4in/renv-system-library

 P ── Loaded and on-disk path mismatch.
`
